Add preliminary TLS1.3 PSK-mode implementation.

rfk opened the pull request:
Fixes https://github.com/mozilla/fxa-pairing-channel/pull/24

(This currently targets https://github.com/mozilla/fxa-pairing-channel/pull/23 but I will rebase against master once that's merged)

This adds actually-appears-to-be-functioning code to implement the TLS1.3 PSK mode. It needs a lot of tests, but I have confirmed that it interoperates successfully with an existing TLS library (tlslite-ng) so I have some base level of confidence that it's doing the right thing in the happy case.

@vladikoff @eoger please try building against this branch and check if things work OK. I will rebase this branch as I add additional testcases and cleanups.

Things still to do:
* [x] Implement the basics of the "alert" protocol, for cleanly closing on crypto errors etc.
* [x] Just, like, lots more tests in general.
* [x] More thorough interop tests with tlslite-ng.
* [ ] ~Look in the NSS codebase for specific testcases to duplicate here.~

vladikoff commented:
First look at this, we cannot make this work yet:

<img src=https://user-images.githubusercontent.com/128755/51409018-9beda880-1b2e-11e9-9226-d46b54feefbb.png width=350 />

vladikoff commented:
Ok we got it to work, it was a problem with psk

rfk commented:
@vladikoff @eoger thanks for the fixes; heads-up that I'm going to port some of them to master and squash the others with my latest work.

eoger commented:
Could we rebase this on `master`?

rfk commented:
Rebased.

martinthomson commented on src/utils.js:
```diff
@@ -66,9 +76,342 @@ export function bytesAreEqual(v1, v2) {
+
+  seek(pos) {
+    this._pos = pos;
+    assert(this._pos <= this.length(), 'do not seek past end of buffer');
```
Shouldn't you assert on < 0 as well?  And why not check the input before applying it?

martinthomson commented on src/utils.js:
```diff
@@ -66,9 +76,342 @@ export function bytesAreEqual(v1, v2) {
+    this._pos = 0;
+  }
+
+  resize(size) {
```
If you can't shrink, then grow() might be the name you want.

martinthomson commented on src/utils.js:
```diff
@@ -66,9 +76,342 @@ export function bytesAreEqual(v1, v2) {
+    return slice;
+  }
+
+  readRemainingBytes() {
```
In TLS, this is probably a hazard.  You should always know how many bytes you are reading.

rfk replied:
👍 ; from a quick grep it looks like this is actually a leftover that is no longer used, so I guess I refactored away whatever hazard was using it...

martinthomson commented on src/utils.js:
```diff
@@ -66,9 +76,342 @@ export function bytesAreEqual(v1, v2) {
+  _readVector(length, cb) {
+    const limit = this.tell() + length;
+    // Keep calling the callback until we've consumed the expected number of bytes.
+    // It can return `true` to indicate we should stop iterating and skip the remainder.
```
I would disable this feature.  The callback should be run once for each item in the list until it is done.

One thing I would do with the callback is pass it a different instance of this object that offers a view onto the same buffer, but with a the shorter length.  Then they can't read past "length" by design and you don't have to worry about inner reads going outside of the space that is available.

martinthomson commented on src/utils.js:
```diff
@@ -66,9 +76,342 @@ export function bytesAreEqual(v1, v2) {
+  //
+
+  _readVector(length, cb) {
+    const limit = this.tell() + length;
```
assert that `limit <= this.length()`

martinthomson commented on src/utils.js:
```diff
@@ -66,9 +76,342 @@ export function bytesAreEqual(v1, v2) {
+  let mismatch = 0;
   for (let i = 0; i < v1.length; i++) {
-    mismatch &= v1[i] !== v2[i];
+    mismatch |= v1[i] ^ v2[i];
```
As we discussed, crypto.subtle.verify might be what you want here.

martinthomson commented on test/tlsconnection.js:
```diff
@@ -0,0 +1,298 @@
+    // XXX TODO: feed it the pre-generated ServerHello.
+  });
+});
+
```
Some tests to consider, though I'm guessing that you have some of these already:

- test reading and writing more thoroughly, including short reads and long reads
- skipping records fails
- damage the ciphertext and the recipient explodes
- skipping handshake messages fails
- missing extensions means failure
- extra extensions (in ClientHello are ignored, in ServerHello generate alerts)
- test that writing blocks/fails prior to the handshake being complete (you could write at the server in 0.5-RTT, but please disable that and test that it is disabled)
- test that you ignore ChangeCipherSpec (or that you choke cleanly) and limit it to during the handshake if you do accept it, if you do ignore it, the server should probably emit one if the client sends a non-zero session ID
- test that you are actually encrypting (yep, that's somewhat tricky, but it's pretty important)
- test without a mocked RNG and see that you get different ClientHello/ServerHello when you run the handshake multiple times
- check that reordered extensions still work fine
- add dummy versions, ciphersuites, and PSK modes to extensions and see that these are ignored
- add a non-zero compression mode and the server should explode
- validate that the version in the record header is correct (TLS 1.0 I think is right: 0x0301)
- validate the version in the ServerHello is TLS 1.2 (0x0303)
- handle padded records properly
- abort if a record only contains zeros (i.e., it is over-padded)
- abort if a handshake record is empty (i.e., it is properly padded with a handshake type, but no data)
- abort if an alert of any kind is received

rfk replied:
> you could write at the server in 0.5-RTT

To clarify, do you mean the server sending application data after it has sent its `Finished`, but before it has received `Finished` from the client in return?

martinthomson replied:
Right.  In TLS 1.3, the server can send before it receives the Finished from the client, but that is generally considered risky in mutually authenticated modes.  For your purposes, you want to wait until you have verified the peer's Finished, in both directions.

vladikoff commented on src/keyschedule.js:
```diff
@@ -0,0 +1,96 @@
+    this.serverApplicationTrafficSecret = null;
+  }
+
+  async addPSK(psk) {
```
can this be `psk = zeros(HASH_LENGTH)` ?

vladikoff commented on src/keyschedule.js:
```diff
@@ -0,0 +1,96 @@
+    this.extBinderKey = await this.deriveSecret('ext binder', EMPTY);
+  }
+
+  async addECDHE(ecdhe) {
```
can this be `ecdhe  = zeros(HASH_LENGTH)`?

rfk replied:
I would prefer to make the caller pass in an explicit "I don't have an ECDHE output" indicator, seems safer to me. I think I'll make this a more explicit check for `null` rather than just falsiness.

vladikoff commented on src/messages.js:
```diff
@@ -0,0 +1,429 @@
+  }
+
+  static fromBytes(bytes) {
+    // Each handshake messages has a type and length prefix, per
```
`messages` -> `message` ?

vladikoff commented on src/messages.js:
```diff
@@ -0,0 +1,429 @@
+  }
+
+  write(buf) {
+    // Each handshake messages has a type and length prefix, per
```
`messages` -> `message`?

vladikoff commented on src/recordlayer.js:
```diff
@@ -0,0 +1,324 @@
+    // * xor with the provided iv
+    const nonce = new Uint8Array(IV_LENGTH);
+    // Our sequence numbers are always less than 2^24, so fit in a Uint32.
+    (new DataView(nonce.buffer)).setUint32(IV_LENGTH - 4, this.seqnum);
```
`new DataView(nonce.buffer).setUint32(IV_LENGTH - 4, this.seqnum);` ? remove extra `(`?

vladikoff commented on src/recordlayer.js:
```diff
@@ -0,0 +1,324 @@
+// The main RecordLayer class.
+
+export class RecordLayer {
+  constructor(sendCallback) {
```
do you want to assert that `sendCallback` is set?

rfk commented:
By way of a brief update here, I'm working my way through the recommended testcases and expect to have them wrapped by in the next day or two.  I also managed to get the test runner to emit coverage reports and run them automatically on CircleCI, latest one from the build output here:

  https://87-160715051-gh.circle-artifacts.com/0/home/circleci/project/coverage/Firefox%2064.0.0%20%28Linux%200.0.0%29/src/index.html

This does a pretty good job of highlighting the error cases that are not yet covered by tests!

@vladikoff, whenever you have a few moments, I'd appreciate your feedback on [this commit](a0ada0e259c76431d9b8d3f837c0c97e4e2e3829) which adds the test coverage reporting. I had to add a third build configuration with `babel-plugin-instanbul` in order to get coverage reporting working with the combination of async/await, babel and karma, but maybe you know of a better approach?

rfk commented:
Ugh, this one test "FxAccountsPairingChannel a Client and Server connected together should communicate successfully if they have the same PSK FAILED" is perma-failing due to a timeout, I wonder if there's some timing issue in setting up the websocket listeners or something, because it passes reliably on my machine...

rfk commented on src/utils.js:
```diff
@@ -24,11 +23,23 @@ export function assert(cond, msg) {
+  // Using `value instanceof Uint8Array` seems to fail in Firefox chrome code
+  // for inscrutable reasons, so we do a less direct check.
+  assert(ArrayBuffer.isView(value), msg);
+  assert(value.BYTES_PER_ELEMENT === 1, msg);
```
@vladikoff do either or both of these pass in Firefox chrome code, or do we need to keep looking for a suitable approach here?

rfk commented:
@martinthomson I've pushed a bunch of tests for the client side of the handshake, and I'm pretty happy with the level of test coverage overall; if you're interested you can view the built coverage report here:

  https://94-160715051-gh.circle-artifacts.com/0/home/circleci/project/coverage/Firefox%2064.0.0%20%28Linux%200.0.0%29/src/index.html

I don't have anything else I plan to add here in response to your first round of comments.

martinthomson commented on src/utils.js:
```diff
@@ -62,13 +73,339 @@ export function utf8ToBytes(str) {
+  }
+
+  writeVector8(cb) {
+    return this._writeVector(Math.pow(2, 8), this.writeUint8.bind(this), cb);
```
is `len => this.writeUint8(len)` better than using bind()?

martinthomson commented on test/FxAccountsPairingChannel.js:
```diff
@@ -23,44 +57,65 @@ describe('FxAccountsPairingChannel', () => {
-      const client = await FxAccountsPairingChannel.InsecurePairingChannel.connect(CHANNEL_SERVER, channelId, channelKey);
-      client.addEventListener('message', ({detail: {data}}) => {
-        CLIENT_RECV.push(data.msg);
+      const client = await FxAccountsPairingChannel.PairingChannel.connect(CHANNEL_SERVER, channelId, channelKey);
```
You run against a network-based service in test?  That seems brittle.  Can you run this server locally? Or at least a cut-down version of the same.

martinthomson commented on test/helpers.js:
```diff
@@ -0,0 +1,254 @@
+    const tampered = bytes.slice();
+    const i = Math.floor(Math.random() * (tampered.byteLength - start) + start);
+    tampered[i] += 1 + Math.floor(Math.random() * 255);
+    return tampered;
```
The problem with this function is that if it does trigger an error, then you won't ever know what conditions produced that error.  If you are going to fuzz, record the way in which your tampering was applied so that you can reproduce problems easily.

martinthomson commented on test/helpers.js:
```diff
@@ -0,0 +1,254 @@
+    const buf = new BufferWriter();
+    buf.writeUint8(1);
+    buf.writeVector24(buf => {
+      buf.writeUint16(typeof opts.version !== 'undefined' ? opts.version : 0x0303);
```
You use `typeof opts.x !== 'undefined' ? opts.x : blah` here, but the simpler `opts.x || blah` above.  Any particular reason?

rfk replied:
IIRC the simpler `opts.x || default` form is used when the default is falsy, and hence there's no reason for the caller to want to pass in an explicit falsy value.  Proably cleaner to just use `typeof opts.x !== 'undefined'` consistently throughout.

martinthomson commented on test/helpers.js:
```diff
@@ -0,0 +1,254 @@
+  signClientHelloMessage: async function (clientHello, psk) {
+    const keyschedule = new KeySchedule();
+    await keyschedule.addPSK(psk);
+    const binder = await keyschedule.calculateFinishedMAC(keyschedule.extBinderKey, clientHello.slice(0, -35));
```
-35 = magic number

rfk commented on src/index.js:
```diff
@@ -72,7 +72,7 @@ export class InsecurePairingChannel extends EventTarget {
           const connection = await ConnectionClass.create(psk, pskId, data => {
             // To send data over the websocket, it needs to be encoded as a safe string.
-            socket.send(bytesToHex(data));
+            socket.send(bytesToBase64url(data));
```
This is an implementation detail of the server we're talking to, rather than of websockets in general.

rfk commented on src/messages.js:
```diff
@@ -0,0 +1,475 @@
+    }
+    if (! seenExtensions.has(EXTENSION_TYPE.PSK_KEY_EXCHANGE_MODES)) {
+      throw new TLSError(ALERT_DESCRIPTION.MISSING_EXTENSION);
+    }
```
Good idea; separating out the logic for each extension would also probably help clean up some of the `testHelper` functions which duplicate parts of this logic.

rfk commented on src/states.js:
```diff
@@ -0,0 +1,370 @@
+// We only support a single PSK, so it's the length of the hash plus one
+// for rendering it as a variable-length byte array, plus two for rendering
+// the variable-length list of PSK binders.
+const PSK_BINDERS_SIZE = HASH_LENGTH + 1 + 2;
```
Yeah, I originally *was* using it on the server, only uncovered after trying to add tests for various badly-formed ClientHello records.

rfk commented on src/recordlayer.js:
```diff
@@ -0,0 +1,329 @@
+      throw new TLSError(ALERT_DESCRIPTION.INTERNAL_ERROR);
+    }
+    // Forbid sending plaintext application data.
+    if (this._sendCipherState.key === null && type === RECORD_TYPE.APPLICATION_DATA) {
```
Note to self: there's a corresponding test to add here as well, that application-data records cannot be sent or received in the handshake phase.

rfk replied:
So, with the change to avoid buffering sent data during the handshake, the `TLSConnection.state` objects handle this state-management already.  For example all the handshake states will now throw if `recvApplicationData` or `sendApplicationData` is called on that state.  Is there still value in duplicating such checks in the record layer?

rfk commented on src/states.js:
```diff
@@ -0,0 +1,370 @@
+        this.conn._closeForRecv();
+        break;
+      default:
+        return await this.handleError(err);
```
Note to self: this won't actually throw the error, but it should!

martinthomson replied:
Maybe you should name the function `error` when you have it throw.

rfk replied:
yep; in my head it's something like `handleAndRethrowError` to make it super clear what will happen.

rfk commented on src/utils.js:
```diff
@@ -62,13 +73,339 @@ export function utf8ToBytes(str) {
 
+export function bytesToBase64url(bytes) {
+  const charCodes = String.fromCharCode.apply(String, bytes);
+  return btoa(charCodes).replace(/\+/g, '-').replace(/\//g, '_');
```
We don't use it on secrets directly, but calling code well might.

rfk replied:
I don't have a good constant-time approach to use here yet, will investigate further...

martinthomson commented on test/tlsconnection.js:
```diff
@@ -0,0 +1,1160 @@
+
+        it('rejects an out-of-order close notification', async () => {
+          await assert.throwsAsync(async () => {
+            await server.recv(TEST_VECTORS.CLIENT_CLOSE);
```
I have to admit to having trouble following the events that lead to this point.  I think that you need a comment explaining where you are and why this fails.

martinthomson commented on test/tlsconnection.js:
```diff
@@ -0,0 +1,1160 @@
+        });
+      });
+
+      it('errors if it receives another handshake message', async () => {
```
This is fine for the server, but not the client.  Right now, you need to accept the NewSessionTicket message, because - unlike in earlier versions of TLS - support for NewSessionTicket isn't negotiated and so servers can send it without prompting.  You can either validate the message and ignore it, or just ignore it, but you do need something.

martinthomson commented on test/tlsconnection.js:
```diff
@@ -0,0 +1,1160 @@
+    }, TLSError, 'UNEXPECTED_MESSAGE');
+  });
+
+  it('accepts a ClientHello advertizing lots of TLS1.3 features that we do not support', async () => {
```
```suggestion
  it('accepts a ClientHello advertising lots of TLS1.3 features that we do not support', async () => {
```

martinthomson commented on test/tlsconnection.js:
```diff
@@ -0,0 +1,1160 @@
+  });
+
+  it('calls crypto.getRandomValues to generate a strong random salt and session id', async () => {
+    assert.equal(crypto.getRandomValues.callCount, 2);
```
Note that getRandomValues is not required for the session ID.  NSS has an awful kludge for that that doesn't involve looking at the random.

I don't mind a little white-box testing.  But you might want to check that both calls ask for 32 octets.

martinthomson commented on test/tlsconnection.js:
```diff
@@ -0,0 +1,1160 @@
+              }, TLSError, 'CLOSE_NOTIFY');
+            });
+
+            describe('is able to send an explicit close in return, and then', () => {
```
You don't seem to have a separate case for server.close() without having first received a close.

martinthomson commented on test/tlsconnection.js:
```diff
@@ -0,0 +1,1160 @@
+              }, TLSError, 'CLOSE_NOTIFY');
+            });
+
+            describe('is able to send an explicit close in return, and then', () => {
```
can you test for sending the close without having received the close first?

martinthomson commented on test/tlsconnection.js:
```diff
@@ -0,0 +1,1160 @@
+      content: await testHelpers.makeServerHelloMessage({
+        extensions: [
+          testHelpers.makePreSharedKeyExtension(1)
+        ]
```
Maybe add version: 0x0304 to this, maybe as a supplementary test, so that you catch the case where the version field in the SH is used to drive version negotiation.  (Not that your code is at all vulnerable to this currently, but it's a good paranoia check.)

rfk replied:
To be clear, the handshake should *fail* if `SH.version` is `0x0304`, right?

martinthomson replied:
Correct.  Fail.

martinthomson commented on test/tlsconnection.js:
```diff
@@ -0,0 +1,1160 @@
+    }, TLSError, 'DECODE_ERROR');
+  });
+
+  it('errors if ServerHello and EncryptedExtensions appear in the same record', async () => {
```
Do you have a test for a partial record being present after the ServerHello or Finished?  I see that you have the code for checking this, but I can't find the test.

rfk replied:
No, I don't believe I do - perhaps that check is being exercised accidentally by some other test.  I'll add one.

martinthomson commented on test/tlsconnection.js:
```diff
@@ -0,0 +1,1160 @@
+      server._keyschedule.serverApplicationTrafficSecret,
+      client._keyschedule.serverApplicationTrafficSecret
+    ));
+  });
```
Can you add tests here for the promise I talked about.

Also, can you add a test that includes calls to send().  It would be good if you could also test that you can close in one direction then continue sending and receiving in the other direction.

rfk commented on test/recordlayer.js:
```diff
@@ -0,0 +1,387 @@
+        assert.equal(SENT_DATA[0][4], 11 + 1 + 16);
+        const ciphertext = SENT_DATA[0].slice(5);
+        assert.equal(ciphertext.byteLength, 11 + 1 + 16);
+        assert.equal(bytesToUtf8(await decryptInnerPlaintext(SENT_DATA[0])), 'hello world' + String.fromCharCode(22));
```
Yeah; I might try doing it like `[type, payload] = decryptInnerPlaintext(...)` and see if that comes out cleaner overall.

rfk commented on src/alerts.js:
```diff
@@ -0,0 +1,64 @@
+    switch (description) {
+      case ALERT_DESCRIPTION.CLOSE_NOTIFY:
+      case ALERT_DESCRIPTION.USER_CANCELLED:
+        this.level = ALERT_LEVEL.WARNING;
```
I'm not using it, and only included it because it was mentioned alongside `close_notify` in [1].  I'll be very happy to remove it.

[1] https://tools.ietf.org/html/rfc8446#section-6.1

rfk commented on src/crypto.js:
```diff
@@ -0,0 +1,117 @@
+export async function decrypt(key, iv, ciphertext, additionalData) {
+  let plaintext;
+  try {
+    plaintext = await crypto.subtle.decrypt({
```
Nope, I think I might have just had it outside for some `console.log` debugging at some point

rfk commented on src/tlsconnection.js:
```diff
@@ -0,0 +1,221 @@
+          await this._state.recvChangeCipherSpec(bytes);
+          return null;
+        case RECORD_TYPE.ALERT:
+          await this._state.recvAlertMessage(TLSError.fromBytes(bytes));
```
I intend for this to throw on `close_notify`, and have the calling code detect it. I'll see how it feels if I create a special error class specifically for `TLSClosed`.

rfk commented on test/utils.js:
```diff
@@ -0,0 +1,469 @@
+        assert.equal(contentsBuf.length(), 2);
+        contentsBuf.readUint24();
+      });
+    }, TLSError, /DECODE_ERROR/);
```
This is mostly just using the existing convenience API of `assert.throws`; I could instead check for a specific numeric value of the alert description, if that seems better?

vladikoff commented:
> Running `npm install;npm test` produces changes to checked-in code. That's not cool

In my experience, At this point `npm install` cannot be trusted to properly use package-lock, it just keeps fetching new things. We should strictly instruct to use the `npm ci` command, that one should never produce changes to the checked-in code.

vladikoff commented on src/extensions.js:
```diff
@@ -0,0 +1,266 @@
+//
+// Extension parsing.
+//
+// This file contains some helpers reading/writing the various kinds
```
`some helpers reading/writing` -> `some helpers for reading/writing` ?

vladikoff replied:
also this comment didn't make it to the final bundle, need to regenerate the build

rfk commented:
Thanks for the detailed feedback @martinthomson, I've updated based on your review with the following key points:

* Extension parsing has been moved into a separate file and helper classes.  This involved refactoring a little bit of the logic of what validation checks live where, but I think the end result is much more easily understood in aggregate.
* Addition on the `connected` promise to resolve/reject on handshake completion, and blocking of sends until it resolves.
* Addition of an explicit `TLSCloseNotify` exception to make it easier to detect close, and some modifications of the websocket wrapper to deal with it appropriately.
* More careful handling of change_cipher_spec, and support for cleanly ignoring NewSessionTicket.
* Split `CipherState into independent `DecryptionState` and `EncryptionState` classes
* The various extra tests you suggested

The diff from the version you reviewed is [here](https://99-160715051-gh.circle-artifacts.com/0/home/circleci/project/coverage/Firefox%2064.0.0%20%28Linux%200.0.0%29/src/index.html), and the latest coverage report is [here](https://99-160715051-gh.circle-artifacts.com/0/home/circleci/project/coverage/Firefox%2064.0.0%20%28Linux%200.0.0%29/src/index.html) for completeness.

vladikoff commented on .circleci/config.yml:
```diff
@@ -21,6 +21,9 @@ jobs:
             npm install
```
let's try to use `npm ci` here

vladikoff commented on README.md:
```diff
@@ -8,19 +8,15 @@ It will be used by the Firefox Accounts pairing flow, with one side
 of the channel being web content from https://accounts.firefox.com and
```
let's add a doc about running `npm ci` and `npm run build` to ensure the build is consistent and no new sub-deps show up. Maybe we can do it as a separate PR.

rfk replied:
Maybe this should be a pre-commit hook or something?  (TBH it kind of weirds me out checking in the build output on every commit, and I wonder if we can only do it on releases or something; but if we're going to do it we should be able to automate it)

martinthomson replied:
pre-commit hooks are local only, you want something in CI ideally

vladikoff commented on src/states.js:
```diff
@@ -0,0 +1,416 @@
+  async sendApplicationData(bytes) {
+    throw this.error;
+  }
+  async sendAlertMessage(err) {
```
is this still used , needed ? I think we use  `_sendAlertMessage` in other places

vladikoff commented on src/index.js:
```diff
@@ -104,23 +104,45 @@ export class InsecurePairingChannel extends EventTarget {
+      if (! this._peerClosed) {
+        this.dispatchEvent(new CustomEvent('error', {
+          detail: {
+            error: new Error('WebSocket unexepectedly closed'),
```
`unexepectedly` --> `unexpectedly`

vladikoff commented on test/recordlayer.js:
```diff
@@ -0,0 +1,396 @@
+        assert.equal(rl._recvDecryptState, null);
+      });
+
+      it('will send encypted handshake records', async () => {
```
`encrypted`

vladikoff commented on test/recordlayer.js:
```diff
@@ -0,0 +1,396 @@
+        assert.equal(type, 22);
+      });
+
+      it('will send encypted application data records', async () => {
```
`encrypted`

vladikoff commented on src/recordlayer.js:
```diff
@@ -0,0 +1,343 @@
+    if (this._recvError !== null) {
+      throw this._recvError;
+    }
+    // For simplicity, we assume that the given data conatins exactly one record.
```
`contains` ?

vladikoff commented on src/states.js:
```diff
@@ -0,0 +1,416 @@
+    const keyschedule = this.conn._keyschedule;
+    const transcript = keyschedule.getTranscript();
+    // Calculate size occupied by the PSK binders.
+    let pskBindersSize = 2; // Vector16 represntation overhead.
```
`representation`  ?

vladikoff commented on test/tlsconnection.js:
```diff
@@ -0,0 +1,1539 @@
+    // keyshares, optional extensions, and other exciting features that we do not support.
+    const data = await server.recv(TEST_VECTORS.EXTENDED_CLIENT_HELLO);
+    assert.equal(data, null);
+    // It sends ServerHello, ChangeCipherSpec, EncyptedExtensions+Finished in response.
```
EncryptedExtensions ?

martinthomson commented on test/tlsconnection.js:
```diff
@@ -780,10 +888,20 @@ describe('the ClientConnection class', () => {
 
+      it('accepts and ignores a NewSessionTicket message', async () => {
+        const encryptor = await testHelpers.makeEncryptionState(client._keyschedule.serverApplicationTrafficSecret, 0);
+        const newSessionTicket = new NewSessionTicket(0, 0, zeros(0), zeros(0), []);
```
It might pay to feed it a bigger ticket than that.  I'd suggest at least one extension, but none are allowed.  Maybe test that an invalid NST is rejected.

martinthomson commented on test/tlsconnection.js:
```diff
@@ -1119,11 +1461,48 @@ describe('the handshake between ClientConnection and ServerConnection', () => {
+    await client.recv(SERVER_SENT[1]);
+    await client.recv(SERVER_SENT[2]);
+    await server.recv(CLIENT_SENT[1]);
+    await client.connected;
```
You can test that a promise is not fulfilled with Promise.any() if you are interested in testing that this isn't fulfilled earlier.

martinthomson commented on test/tlsconnection.js:
```diff
@@ -1119,11 +1461,48 @@ describe('the handshake between ClientConnection and ServerConnection', () => {
+    const CLIENT_SENT = [], SERVER_SENT = [];
+    const server = await ServerConnection.create(TEST_VECTORS.PSK, TEST_VECTORS.PSK_ID, data => SERVER_SENT.push(data));
+    const client = await ClientConnection.create(TEST_VECTORS.PSK, TEST_VECTORS.PSK_ID, data => CLIENT_SENT.push(data));
+    await server.recv(CLIENT_SENT[0]);
```
// CH

martinthomson commented on test/tlsconnection.js:
```diff
@@ -1119,11 +1461,48 @@ describe('the handshake between ClientConnection and ServerConnection', () => {
+    const server = await ServerConnection.create(TEST_VECTORS.PSK, TEST_VECTORS.PSK_ID, data => SERVER_SENT.push(data));
+    const client = await ClientConnection.create(TEST_VECTORS.PSK, TEST_VECTORS.PSK_ID, data => CLIENT_SENT.push(data));
+    await server.recv(CLIENT_SENT[0]);
+    await client.recv(SERVER_SENT[0]);
```
// SH

martinthomson commented on test/tlsconnection.js:
```diff
@@ -1119,11 +1461,48 @@ describe('the handshake between ClientConnection and ServerConnection', () => {
+    const client = await ClientConnection.create(TEST_VECTORS.PSK, TEST_VECTORS.PSK_ID, data => CLIENT_SENT.push(data));
+    await server.recv(CLIENT_SENT[0]);
+    await client.recv(SERVER_SENT[0]);
+    await client.recv(SERVER_SENT[1]);
```
// EE+Finished

rfk replied:
I'll add comments, but for completeness, this is ChangeCipherSpec...

martinthomson commented on test/tlsconnection.js:
```diff
@@ -1119,11 +1461,48 @@ describe('the handshake between ClientConnection and ServerConnection', () => {
+    await server.recv(CLIENT_SENT[0]);
+    await client.recv(SERVER_SENT[0]);
+    await client.recv(SERVER_SENT[1]);
+    await client.recv(SERVER_SENT[2]);
```
??  Finished?

NSS bundles everything together, so I'm unused to multiple messages.

rfk replied:
...and this is EE+Finished.

martinthomson commented on test/tlsconnection.js:
```diff
@@ -1119,11 +1461,48 @@ describe('the handshake between ClientConnection and ServerConnection', () => {
+    await client.recv(SERVER_SENT[0]);
+    await client.recv(SERVER_SENT[1]);
+    await client.recv(SERVER_SENT[2]);
+    await server.recv(CLIENT_SENT[1]);
```
// Client Finished

rfk commented on src/states.js:
```diff
@@ -52,41 +49,38 @@ class State {
-        this.conn._closeForRecv();
-        break;
+        this.conn._closeForRecv(alert);
+        throw alert;
```
Basically, yes.  I think a better approach is probably to shuffle the API to use an `EventEmitter` style, so it can emit a `close` event in this case, which is basically what the higher-level `PairingChannel` abstraction does anyway.  But it was a bigger refactor than I wanted to take on inline in this PR.

rfk commented:
Thanks so much @martinthomson!
